File: app/src/main/java/com/amplifyframework/datastore/sample/LogWindow.java
package com.amplifyframework.datastore.sample;

import android.graphics.Typeface;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.IdRes;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

final class LogWindow {
    private LogLineAdapter adapter;
    private RecyclerView recyclerView;

    LogWindow(ViewGroup parent, @IdRes int resourceId) {
        this.adapter = new LogLineAdapter();

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(parent.getContext());
        linearLayoutManager.setStackFromEnd(true);

        recyclerView = parent.findViewById(resourceId);
        recyclerView.setLayoutManager(linearLayoutManager);
        recyclerView.setAdapter(adapter);
    }

    void add(LogLine logLine) {
        adapter.add(logLine);
        recyclerView.scrollToPosition(this.adapter.size() - 1);
    }

    void clear() {
        adapter.clear();
    }

    private static final class LogLineAdapter extends RecyclerView.Adapter<LogLineAdapter.LineItemViewHolder> {
        private final List<LogLine> lines;

        LogLineAdapter() {
            this.lines = new ArrayList<>();
        }

        void add(LogLine line) {
            lines.add(line);
            super.notifyDataSetChanged();
        }

        void clear() {
            lines.clear();
            super.notifyDataSetChanged();
        }

        int size() {
            return lines.size();
        }

        @NonNull
        @Override
        public LineItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            RelativeLayout lintItemView = (RelativeLayout) LayoutInflater.from(parent.getContext())
                .inflate(R.layout.line_item, parent, false);
            return new LineItemViewHolder(lintItemView);
        }

        @Override
        public void onBindViewHolder(@NonNull LineItemViewHolder holder, int position) {
            LogLine line = lines.get(position);
            holder.setLine(line);
        }

        @Override
        public int getItemCount() {
            return lines.size();
        }

        private static class LineItemViewHolder extends RecyclerView.ViewHolder {
            private final TextView titleView;
            private final TextView detailsView;
            private final TextView typeView;

            LineItemViewHolder(RelativeLayout lintItemView) {
                super(lintItemView);
                titleView = lintItemView.findViewById(R.id.title);
                titleView.setTypeface(null, Typeface.BOLD);

                detailsView = lintItemView.findViewById(R.id.details);
                typeView = lintItemView.findViewById(R.id.type);
            }

            void setLine(LogLine line) {
                titleView.setText(line.getTitle());
                detailsView.setText(line.getDetails());
                typeView.setText(line.getEventType().getText());
            }
        }
    }
}

File: app/src/main/java/com/amplifyframework/datastore/sample/LogWindow.kt
package com.amplifyframework.datastore.sample

import android.graphics.Typeface
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.RelativeLayout
import android.widget.TextView

import androidx.annotation.IdRes
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.Adapter
import androidx.recyclerview.widget.RecyclerView.ViewHolder

import com.amplifyframework.datastore.sample.LogWindow.LogLineAdapter.LineItemViewHolder
import com.amplifyframework.datastore.sample.R.id
import com.amplifyframework.datastore.sample.R.layout

internal class LogWindow(parent: ViewGroup, @IdRes resourceId: Int) {
    private val adapter: LogLineAdapter = LogLineAdapter()
    private val recyclerView: RecyclerView
    init {
        val linearLayoutManager = LinearLayoutManager(parent.context)
        linearLayoutManager.stackFromEnd = true
        recyclerView = parent.findViewById(resourceId)
        recyclerView.layoutManager = linearLayoutManager
        recyclerView.adapter = adapter
    }

    fun add(logLine: LogLine) {
        adapter.add(logLine)
        recyclerView.scrollToPosition(adapter.size() - 1)
    }

    fun clear(): Unit = adapter.clear()

    private class LogLineAdapter internal constructor() : Adapter<LineItemViewHolder>() {
        private val lines: MutableList<LogLine> = ArrayList()

        fun add(line: LogLine) {
            lines.add(line)
            super.notifyDataSetChanged()
        }

        fun clear() {
            lines.clear()
            super.notifyDataSetChanged()
        }

        fun size(): Int = lines.size

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LineItemViewHolder {
            val lintItemView = LayoutInflater.from(parent.context)
                .inflate(layout.line_item, parent, false) as RelativeLayout
            return LineItemViewHolder(lintItemView)
        }

        override fun onBindViewHolder(holder: LineItemViewHolder, position: Int) = holder.setLine(lines[position])

        override fun getItemCount(): Int = lines.size

        private class LineItemViewHolder internal constructor(lintItemView: RelativeLayout) : ViewHolder(lintItemView) {
            private val detailsView: TextView = lintItemView.findViewById(id.details)
            private val typeView: TextView = lintItemView.findViewById(id.type)
            private val titleView: TextView = lintItemView.findViewById(id.title)
            init {
                titleView.setTypeface(null, Typeface.BOLD)
            }

            fun setLine(line: LogLine) {
                titleView.text = line.title
                detailsView.text = line.details
                typeView.text = line.eventType.text
            }
        }
    }
}
